src/sentry/middleware/integrations/parsers/base.py

```diff
@@ -0,0 +1,143 @@
+from __future__ import annotations
+
+import abc
+import logging
+from concurrent.futures import ThreadPoolExecutor, as_completed
+from typing import Callable, Iterable, Sequence
+
+from django.http.request import HttpRequest
+from django.urls import ResolverMatch, resolve
+
+from sentry.models.integrations import Integration, OrganizationIntegration
+from sentry.services.hybrid_cloud.organization import ApiOrganizationSummary, organization_service
+from sentry.silo import SiloLimit, SiloMode
+from sentry.silo.client import RegionSiloClient
+from sentry.types.region import Region, get_region_for_organization
+from sentry.utils.sdk import capture_exception
+
+logger = logging.getLogger(__name__)
+
+
+class BaseRequestParser(abc.ABC):
+    """Base Class for Integration Request Parsers"""
+
+    def __init__(self, request: HttpRequest, response_handler: Callable):
+        self.request = request
+        self.match: ResolverMatch = resolve(self.request.path)
+        self.response_handler = response_handler
+        self.error_message = "Integration Request Parsers should only be run on the control silo."
+
+    def _ensure_control_silo(self):
+        if SiloMode.get_current_mode() != SiloMode.CONTROL:
+            logger.error(
+                "request_parser.base.silo_error",
+                extra={"path": self.request.path, "silo": SiloMode.get_current_mode()},
+            )
+            raise SiloLimit.AvailabilityError(self.error_message)
+
+    def get_response_from_control_silo(self):
+        """
+        Used to handle the request directly on the control silo.
+        """
+        self._ensure_control_silo()
+        return self.response_handler(self.request)
+
+    def _get_response_from_region_silo(self, region: Region):
+        region_client = RegionSiloClient(region)
+        return region_client.proxy_request(self.request).to_http_response()
+
+    def get_response_from_region_silos(self, regions: Iterable[Region]):
```

Comment:
In general this seems risky.  What are the deliverability requirements that we should have for this method?  As written, some silos might get the message, and others don't, and the caller has no idea what sort of success occurs (partial, complete, etc).

It seems certain that if we must be sending to multiple regions we'll need to track the success state of those calls, OR fail this call outright unless every region receives the message.

Reply:
That is a good callout, I think you're right. I'll rework the function signature to return a list of responses, or mapping from region to response so the implenmenting classes can handle decide what to do in the event of partial failures.